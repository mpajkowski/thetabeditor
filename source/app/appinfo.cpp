/*
 * Copyright (C) 2014 Cameron White
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "appinfo.h"

#include <QCoreApplication>
#include <QtGlobal>

namespace AppInfo
{
const char *BUG_TRACKER_URL =
    "https://github.com/powertab/thetabeditor/issues/";
const char *APPLICATION_NAME = "The Tab Editor";
const char *APPLICATION_VERSION = "2.0";

#if defined(Q_OS_WIN)
const char *ORGANIZATION_NAME = "Power Tab";
const char *APPLICATION_ID = "The Tab Editor";
#else
const char *ORGANIZATION_NAME = "powertab";
const char *APPLICATION_ID = "thetabeditor";
#endif

std::string getAbsolutePath(const char *relative_path)
{
    std::string path = QCoreApplication::applicationDirPath().toStdString();
    path += "/";
    path += relative_path;
    return path;
}
} // namespace AppInfo
