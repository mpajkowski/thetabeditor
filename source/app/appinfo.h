/*
 * Copyright (C) 2014 Cameron White
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef APP_APPINFO_H
#define APP_APPINFO_H

#include <string>

namespace AppInfo {
extern const char* BUG_TRACKER_URL;
/// Organization name used by QSettings.
extern const char* ORGANIZATION_NAME;
/// Application name used by QSettings.
extern const char* APPLICATION_ID;
/// Application name that should be displayed to the user.
extern const char* APPLICATION_NAME;
extern const char* APPLICATION_VERSION;

/// Given a path relative to the location of the executable, return the full
/// path.
std::string getAbsolutePath(const char* relative_path);
} // namespace AppInfo

#endif
