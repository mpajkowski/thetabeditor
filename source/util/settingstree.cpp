/*
 * Copyright (C) 2015 Cameron White
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "settingstree.h"

#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>

#include <rapidjson/document.h>
#include <rapidjson/error/en.h>
#include <rapidjson/prettywriter.h>
#include <util/rapidjson_iostreams.h>

using SettingValue = SettingsTree::SettingValue;
using SettingList = SettingsTree::SettingList;
using SettingMap = SettingsTree::SettingMap;

struct IsMap : public boost::static_visitor<bool>
{
  template<typename T>
  bool operator()(T const&) const
  {
    return false;
  }

  bool operator()(SettingMap const&) const { return true; }
};

struct Inserter : public boost::static_visitor<>
{
  Inserter(const std::string& key, SettingValue value)
    : myIndex(0)
    , myNewValue(std::move(value))
  {
    boost::algorithm::split(myComponents, key, boost::is_any_of("/"));
  }

  template<typename T>
  void operator()(T&)
  {
    assert(false);
  }

  void operator()(SettingMap& map)
  {
    SettingValue& value = map[myComponents[myIndex]];

    if (myIndex == myComponents.size() - 1)
      value = myNewValue;
    else {
      if (!boost::apply_visitor(IsMap(), value))
        value = SettingMap();

      ++myIndex;
      boost::apply_visitor(*this, value);
    }
  }

private:
  std::vector<std::string> myComponents;
  size_t myIndex;
  SettingValue myNewValue;
};

struct Finder : public boost::static_visitor<boost::optional<SettingValue>>
{
  Finder(const std::string& key)
    : myIndex(0)
  {
    boost::algorithm::split(myComponents, key, boost::is_any_of("/"));
  }

  template<typename T>
  boost::optional<SettingValue> operator()(T const&) const
  {
    return boost::none;
  }

  boost::optional<SettingValue> operator()(SettingMap const& map) const
  {
    auto it = map.find(myComponents[myIndex]);
    if (it == map.end())
      return boost::none;

    SettingValue const& value = it->second;
    if (myIndex == myComponents.size() - 1)
      return value;
    else {
      if (!boost::apply_visitor(IsMap(), value))
        return boost::none;

      ++myIndex;
      return boost::apply_visitor(*this, value);
    }
  }

private:
  std::vector<std::string> myComponents;
  mutable size_t myIndex;
};

struct Remover : public boost::static_visitor<void>
{
  Remover(const std::string& key)
    : myIndex(0)
  {
    boost::algorithm::split(myComponents, key, boost::is_any_of("/"));
  }

  template<typename T>
  void operator()(T&)
  {}

  void operator()(SettingMap& map)
  {
    auto it = map.find(myComponents[myIndex]);
    if (it != map.end()) {
      if (myIndex == myComponents.size() - 1)
        map.erase(it);
      else {
        ++myIndex;
        boost::apply_visitor(*this, it->second);
      }
    }
  }

private:
  std::vector<std::string> myComponents;
  size_t myIndex;
};

static void parseValue(SettingValue& value, const rapidjson::Value& json_val)
{
  if (json_val.IsInt())
    value = json_val.GetInt();
  else if (json_val.IsString())
    value = std::string(json_val.GetString(), json_val.GetStringLength());
  else if (json_val.IsBool())
    value = json_val.GetBool();
  else if (json_val.IsArray()) {
    SettingList value_list;

    for (auto it = json_val.Begin(); it != json_val.End(); ++it) {
      SettingValue child_val;
      parseValue(child_val, *it);
      value_list.push_back(child_val);
    }

    value = value_list;
  } else if (json_val.IsObject()) {
    SettingMap value_map;

    for (auto it = json_val.MemberBegin(); it != json_val.MemberEnd(); ++it) {
      SettingValue child_val;
      parseValue(child_val, it->value);
      value_map[it->name.GetString()] = child_val;
    }

    value = value_map;
  } else
    throw std::runtime_error("Unexpected JSON value type.");
}

struct JSONSerializer : public boost::static_visitor<void>
{
  JSONSerializer(std::ostream& os)
    : myStream(os)
    , myWriter(myStream)
  {}

  void operator()(int x) { myWriter.Int(x); }

  void operator()(bool x) { myWriter.Bool(x); }

  void operator()(const std::string& s)
  {
    myWriter.String(s.c_str(), static_cast<rapidjson::SizeType>(s.length()));
  }

  void operator()(SettingList const& list)
  {
    myWriter.StartArray();
    for (auto&& value : list)
      boost::apply_visitor(*this, value);
    myWriter.EndArray();
  }

  void operator()(SettingMap const& map)
  {
    myWriter.StartObject();

    // Output in sorted order.
    std::vector<std::string> keys;
    keys.reserve(map.size());
    for (auto&& pair : map)
      keys.push_back(pair.first);

    std::sort(keys.begin(), keys.end());

    for (auto&& key : keys) {
      myWriter.Key(key.c_str());
      boost::apply_visitor(*this, map.at(key));
    }

    myWriter.EndObject();
  }

private:
  Util::RapidJSON::OStreamWrapper myStream;
  rapidjson::PrettyWriter<decltype(myStream)> myWriter;
};

SettingsTree::SettingsTree()
  : myTree(SettingMap())
{}

void SettingsTree::setImpl(const std::string& key, SettingValue const& value)
{
  Inserter inserter(key, value);
  boost::apply_visitor(inserter, myTree);
}

boost::optional<SettingValue> SettingsTree::find(const std::string& key) const
{
  return boost::apply_visitor(Finder(key), myTree);
}

void SettingsTree::remove(const std::string& key)
{
  Remover visitor(key);
  return boost::apply_visitor(visitor, myTree);
}

void SettingsTree::loadFromJSON(std::istream& is)
{
  Util::RapidJSON::IStreamWrapper stream(is);

  rapidjson::Document document;
  document.ParseStream(stream);

  if (document.HasParseError()) {
    throw std::runtime_error("Parse error at offset " + std::to_string(document.GetErrorOffset()) + ": " +
                             GetParseError_En(document.GetParseError()));
  }

  parseValue(myTree, document);
}

void SettingsTree::saveToJSON(std::ostream& os) const
{
  JSONSerializer serializer(os);
  boost::apply_visitor(serializer, myTree);
}
