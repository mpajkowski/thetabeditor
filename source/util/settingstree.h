/*
 * Copyright (C) 2015 Cameron White
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef APP_SETTINGSTREE_H
#define APP_SETTINGSTREE_H

#include <boost/optional/optional.hpp>
#include <boost/variant/get.hpp>
#include <boost/variant/recursive_variant.hpp>
#include <boost/variant/variant.hpp>

#include <iosfwd>
#include <string>
#include <unordered_map>
#include <vector>

template<typename T>
class Setting
{
public:
  Setting(std::string key, T default_value)
    : myKey(std::move(key))
    , myDefaultValue(std::move(default_value))
  {}

  const std::string myKey;
  const T myDefaultValue;
};

class SettingsTree
{
public:
  using SettingValue =
    boost::make_recursive_variant<int,
                                  std::string,
                                  bool,
                                  std::vector<boost::recursive_variant_>,
                                  std::unordered_map<std::string, boost::recursive_variant_>>::type;

  using SettingList = std::vector<SettingValue>;
  using SettingMap = std::unordered_map<std::string, SettingValue>;

  SettingsTree();

  /// Retrieve the value associated with the given key, or the default value
  /// if not found.
  /// @throws std::exception If T does not match the value's type.
  template<typename T>
  T get(const std::string& key, T const& default_val = T()) const;

  /// Retrieve a list of values associated with the given key.
  template<typename T>
  std::vector<T> getList(const std::string& key) const;

  template<typename T>
  T get(const Setting<T>& setting) const
  {
    return get<T>(setting.myKey, setting.myDefaultValue);
  }

  template<typename T>
  std::vector<T> get(const Setting<std::vector<T>>& setting) const
  {
    return getList<T>(setting.myKey);
  }

  /// Set the value associated with the key.
  template<typename T>
  void set(const std::string& key, T const& value);

  /// Since the implicit conversion from const char * to bool takes
  /// precedence over the implicit std::string constructor, add an explicit
  /// overload that does the right thing.
  void set(const std::string& key, const char* value) { set(key, std::string(value)); }

  /// Set the value associated with the key to a list of values.
  template<typename T>
  void setList(const std::string& key, const std::vector<T>& values);

  template<typename T>
  void set(const Setting<T>& setting, T const& value)
  {
    set(setting.myKey, value);
  }

  template<typename T>
  void set(const Setting<std::vector<T>>& setting, const std::vector<T>& value)
  {
    setList(setting.myKey, value);
  }

  /// Removes a setting.
  void remove(const std::string& key);

  void loadFromJSON(std::istream& is);
  void saveToJSON(std::ostream& os) const;

  /// On OSX only, saves the settings using NSUserDefaults.
  void saveToPlist() const;
  void loadFromPlist();

private:
  void setImpl(const std::string& key, SettingValue const& value);
  boost::optional<SettingValue> find(const std::string& key) const;

  SettingValue myTree;
};

/// Specialize to support custom types as setting values.
template<typename T>
struct SettingValueConverter
{
  static SettingsTree::SettingValue to(T const& t) { return t; }

  static T from(const SettingsTree::SettingValue& v) { return boost::get<T>(v); }
};

template<typename T>
T SettingsTree::get(const std::string& key, T const& default_val) const
{
  boost::optional<SettingValue> val = find(key);
  return val ? SettingValueConverter<T>::from(*val) : default_val;
}

template<typename T>
std::vector<T> SettingsTree::getList(const std::string& key) const
{
  auto values = get<std::vector<SettingValue>>(key);

  std::vector<T> ts;
  ts.reserve(values.size());
  for (auto&& val : values)
    ts.push_back(SettingValueConverter<T>::from(val));

  return ts;
}

template<typename T>
void SettingsTree::set(const std::string& key, T const& val)
{
  setImpl(key, SettingValueConverter<T>::to(val));
}

template<typename T>
void SettingsTree::setList(const std::string& key, const std::vector<T>& values)
{
  std::vector<SettingValue> tmp(values.begin(), values.end());
  setImpl(key, SettingValue(tmp));
}

#endif
