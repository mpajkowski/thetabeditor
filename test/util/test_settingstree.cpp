/*
 * Copyright (C) 2015 Cameron White
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <catch.hpp>

#include <app/appinfo.h>
#include <fstream>
#include <string>
#include <sstream>
#include <util/settingstree.h>

static const std::string theKey = "foo/bar";

TEST_CASE("Util/SettingsTree/Int")
{
  SettingsTree settings;

  REQUIRE(settings.get(theKey, -1) == -1);
  settings.set(theKey, 1);
  REQUIRE(settings.get(theKey, -1) == 1);
  settings.set(theKey, 2);
  REQUIRE(settings.get(theKey, -1) == 2);
  REQUIRE_THROWS(settings.get<std::string>(theKey));
}

TEST_CASE("Util/SettingsTree/Bool")
{
  SettingsTree settings;

  REQUIRE(settings.get(theKey, true) == true);
  settings.set(theKey, false);
  REQUIRE(settings.get(theKey, true) == false);
  REQUIRE_THROWS(settings.get<int>(theKey));
}

TEST_CASE("Util/SettingsTree/String")
{
  SettingsTree settings;

  settings.set(theKey, "asdf");
  REQUIRE(settings.get<std::string>(theKey) == "asdf");
  REQUIRE_THROWS(settings.get<int>(theKey, -1));
}

TEST_CASE("Util/SettingsTree/StringList")
{
  SettingsTree settings;

  std::vector<std::string> strings = { "abc", "def" };

  for (int i = 0; i < 3; ++i)
    settings.setList(theKey, strings);

  REQUIRE(settings.getList<std::string>(theKey) == strings);
}

TEST_CASE("Util/SettingsTree/IntList")
{
  SettingsTree settings;

  std::vector<int> vals = { 1, 2, 3 };
  settings.setList(theKey, vals);

  REQUIRE(settings.getList<int>(theKey) == vals);
}

TEST_CASE("Util/SettingsTree/Paths")
{
  SettingsTree settings;

  settings.set("foo", 1);
  REQUIRE(settings.get("foo", -1) == 1);
  settings.set("foo/bar", 2);
  settings.set("foo/baz", "asdf");
  REQUIRE(settings.get("foo/bar", -1) == 2);
  REQUIRE(settings.get<std::string>("foo/baz") == "asdf");
  REQUIRE_THROWS(settings.get("foo", -1));
}

TEST_CASE("Util/SettingsTree/JSON/Export")
{
  SettingsTree settings;

  settings.set("key_a", -123);
  settings.setList("key_b", std::vector<int>({ 1, 2, 3 }));
  settings.setList("key_c", std::vector<std::string>({ "string1", "string2", "string3" }));
  settings.set("parent/child_a", 234);
  settings.set("parent/child_b", "asdf");
  settings.set("parent/child_c", true);
  settings.set("parent/child_d", false);

  std::ostringstream output;
  settings.saveToJSON(output);

  std::ifstream expected_file(AppInfo::getAbsolutePath("data/test_settingstree_expected.json"));
  std::stringstream expected;
  expected << expected_file.rdbuf();

  REQUIRE(output.str() == expected.str());
}

TEST_CASE("Util/SettingsTree/JSON/Import")
{
  std::ifstream input_file(AppInfo::getAbsolutePath("data/test_settingstree_expected.json"));

  SettingsTree settings;
  settings.loadFromJSON(input_file);

  REQUIRE(settings.get<int>("key_a", 0) == -123);
  REQUIRE(settings.getList<int>("key_b") == std::vector<int>({ 1, 2, 3 }));
  REQUIRE(settings.getList<std::string>("key_c") ==
          std::vector<std::string>({ "string1", "string2", "string3" }));
  REQUIRE(settings.get<int>("parent/child_a", 0) == 234);
  REQUIRE(settings.get<std::string>("parent/child_b") == "asdf");
  REQUIRE(settings.get<bool>("parent/child_c") == true);
  REQUIRE(settings.get<bool>("parent/child_d") == false);
}

TEST_CASE("Util/SettingsTree/Remove")
{
  SettingsTree settings;

  settings.set(theKey, 1);
  REQUIRE(settings.get(theKey, -1) == 1);
  settings.remove(theKey);
  REQUIRE(settings.get(theKey, -1) == -1);
}
