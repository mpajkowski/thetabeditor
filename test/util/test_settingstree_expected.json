{
    "key_a": -123,
    "key_b": [
        1,
        2,
        3
    ],
    "key_c": [
        "string1",
        "string2",
        "string3"
    ],
    "parent": {
        "child_a": 234,
        "child_b": "asdf",
        "child_c": true,
        "child_d": false
    }
}